[#42158] Add direct action hover menu

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/42158
- https://community.openproject.org/work_packages/42159
- added hover menu
  - style of the spot link is overwritten, as current style is not in design spots available
- added direct actions
  - open in location
  - unlink

Kharonus committed:
Merge branch 'dev' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] added hover menu and file open actions

Kharonus committed:
[#42158] added hover title on action elements

cbliard committed:
add matrix gem to get the Docker image to build again

matrix gem is no longer included in ruby 3.1 and is needed by the prawn gem, and prawn does not explicitly declare the dependency to the matrix gem in its latest version. We need to explicitly add it.

lindenthal commented:
Hi @wielinde,
the pullpreview instance is failing. I had the same problem in PR10664. There was a dependency missing. Can you please check?
Thanks
Niels

b12f commented:
This commit fixes that @Kharonus https://github.com/opf/openproject/pull/10664/commits/bdcd06fa7dea2c8cb734eefa8bc293fbbf4f1d49

Kharonus committed:
Merge branch 'implementation/42157-list-file-links' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] added minor color changes

Kharonus committed:
Merge branch 'dev' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
Merge branch 'implementation/42157-list-file-links' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] Added remove file link direct action

- added new icons to font
- added remove function to file links repo service
- added remove handler to file link list item

Kharonus committed:
[#42158] add manage file links permission switch

lindenthal commented:
Hi @b12f,
this PR is failing. Can you please check?

Kharonus committed:
Merge branch 'implementation/42157-list-file-links' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] fixed merge failure

Kharonus committed:
[#42158] reworked style of storage information box

Kharonus committed:
[#42158] finalized file list style

Kharonus committed:
[#42158] added styling for headers, action button

Kharonus committed:
[#42158] fixed minor spacings and styling issues

Kharonus committed:
removed negative margin from icons in spot links

Kharonus committed:
[#42158] fixed last styling issues

Commit:
Add compact version of spot-list

This adds the modifier `spot-list_compact`, which compresses `spot-list`
vertically by reducing paddings. It also increases the padding on the
right hand side of items, since it differed from the left.

Commit:
Update spot-list hover color and compact spacings

Kharonus committed:
Merge branch 'dev' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] switched to behaviour subject for stable state

Kharonus committed:
Merge remote-tracking branch 'origin/feature/spot-list_compact' into chore/add-spot-hover-menu

Kharonus committed:
[chore] new hover menu for spot-list

Kharonus committed:
[chore] removed exception and renamed menu

Kharonus committed:
[chore] introduced floating wrapper

- enable correct hover effects on list items with flaoting actions
- align children to top when list item gets line breaks

Kharonus committed:
[chore] moved comment to paragraph

Kharonus committed:
Merge branch 'dev' into chore/add-spot-hover-menu

Kharonus committed:
Merge branch 'chore/add-spot-hover-menu' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] amended spot list styles

Kharonus committed:
[#42158] replaced custom file list with spot list

- with some overwrites
- enables multi content design

apfohl commented on frontend/src/app/core/state/file-links/file-links.service.ts:
```diff
@@ -90,4 +91,36 @@ export class FileLinksResourceService extends ResourceCollectionService<IFileLin
+      .subscribe(() => {
+        applyTransaction(() => {
+          this.store.remove(fileLink.id);
+          this.store.update(({ collections }) => (
```
We have a function for updating collections, havn't we?

Kharonus replied:
extracted to higher level function

Kharonus committed:
Merge branch 'dev' into implementation/42158-add-direct-action-hover-menu

Kharonus committed:
[#42158] adapt new spot-list styles for file list

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.component.ts:
```diff
@@ -37,34 +45,61 @@ import {
+    if (this.fileLink.originData.lastModifiedByName) {
+      this.principalRendererService.render(
+        this.avatar.nativeElement,
+        { name: this.fileLink.originData.lastModifiedByName, href: '/users/1' },
```
Always linking to `users/1` is probably not correct. Since we probably don't know the real href we can perhaps also leave it empty?

Kharonus replied:
thing here is, that the principal renderer service is not hardened against putting a `PrincipalLike` without an href. It just throws an error and exits. In a following branch I detected that misbehaviour and introduced another `PrincipalType`, the `external_user`, which basically omits any gravatar loading.

HDinger replied:
Ah okay, if that is already tackled, it is fine for me to keep it like this for now. Maybe you can add a small comment to explain it, and add a TODO

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.component.ts:
```diff
@@ -37,34 +45,61 @@ import {
     if (this.fileLink.originData.lastModifiedAt) {
-      const date = this.timezoneService.formattedDate(this.fileLink.originData.lastModifiedAt);
-      this.infoTimestampText = this.i18n.t('js.label_modified_at', { date });
+      this.infoTimestampText = this.timezoneService.parseDatetime(this.fileLink.originData.lastModifiedAt).fromNow();
```
`this.fileLink.originData` is accessed quite often. Please consider extracting that to a getter or a variable.

Kharonus replied:
extracted to private getter

HDinger commented on frontend/src/global_styles/content/work_packages/tabs/_files.sass:
```diff
@@ -27,71 +27,75 @@
-
-    &--file-info-timestamp
-      margin-right: 8px
+  .op-files-tab--file-list
```
Please use the `&--file-list` notation instead of `.op-files-tab--file-list`. It is more structural when it comes to implementing BEM styles and a notation we implicitly agreed on.

Kharonus replied:
replaced

Kharonus replied:
if I do that, the selector is not strong enough to set the overwritten paddings on the `spot-list--item-action` element, which is here altered with the class `op-files-tab--file-list-item-action` with left and right padding set to 0.

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.html:
```diff
@@ -1,25 +1,44 @@
-    [textContent]="infoTimestampText"
-  ></span>
+    <span
+      class="op-files-tab--file-list-item-text op-files-tab--color-grey"
```
`op-files-tab--color-grey` is not a good class name because of two reasons:
1. According to BEM "color-grey" would be your element selector. However, it clearly is no element but rather a modifier (?)
2. If it is a modifier, a semantic naming would be better, to know when to use it (e.g `_disabled`,  or `_loading`).

In your case, I guess it depends on the file-type. So instead of passing the colour via the .ts code, please pass any other string referencing the type. Thus, in case we want to change the colour of a file type later on, we only have to do that in one CSS file, instead of multiple places all over the code.

Kharonus replied:
replaced colors with icon classes. Also inline colors for texts (actually only the grey for the date, and the storage information text)

HDinger commented on frontend/src/global_styles/content/work_packages/tabs/_files.sass:
```diff
@@ -27,71 +27,75 @@
+          text-overflow: ellipsis
+
+        .op-files-tab--file-list-item-text
+          font-size: $spot-spacing-0-75
```
As the name suggests, the `spacing` variables are meant to use for spacings (padding, margin, width etc.) Especially for the font-size we do have predefined mixins (which you are already using above). Taken from Figma, I think `spot-caption` is the right one here

Kharonus replied:
replaced some font sizes with include for spot typography mixins.

HDinger reviewed (changes requested):
Just some minor remarks:

* Old powerpoint and excel files (without the `x` at the end) have a different mime type and are thus not recognised as such files:
<img width="188" alt="Bildschirmfoto 2022-06-13 um 14 40 35" src="https://user-images.githubusercontent.com/7457313/173355742-10275724-7707-465d-a422-052c470ad362.png">

* The button to add new files looks weird. I think the styling is missing:
<img width="513" alt="Bildschirmfoto 2022-06-13 um 14 42 27" src="https://user-images.githubusercontent.com/7457313/173356020-6d34601a-c069-48cf-8319-97d4f95bbe61.png">
 
* I don't really get the difference between clicking the first icon, clicking the file name or clicking somewhere in between. It seems like it all does the same. Is that intended?

* Personally I am fan of using sub-folders for each component in code. So `file-links/file-link-list/file-link-list-item/` and `file-links/file-link-list/storage-information/` and so on. Imho that gives a better structure, but that might be only my personal opinion 🤷‍♀️

Kharonus commented:
To answer some questions first:

- we can add more mimetype mappings, should be very simple, which is why I extracted that into a factory function. Any suggestions which ones to add? Or should I asked product team?
- I thought I hit the default button pretty well, at least, it was specified in figma like here: https://www.figma.com/file/gtLQfPe09X7XugAH8L7dTy/Nextcloud-Integration?node-id=936%3A32804. Do you think, we need more styles on it? EDIT: rechecked it, there were some spacings missing, due do the inner spot-icon having no width and height set. fixed.
- first things first: the whole list item is an anchor tag, hence, there is no difference, where to click it, it will always open the file on nextcloud (behaves as specified). The folder icon in the floating actions instead opens the files location. In the nextcloud it will look differently (first things opens the inline viewer on nextcloud, the other opens the nextcloud folder with the wanted file selected. The magic here happens in the URL query param `openfile`, which you can observe in the redirected urls)
- I can structure the code as you wanted. I simply took this structure from the attachment list, but have no pain in putting those in more elaborate folders. But I will do so in a separate PR, as otherwise I will get some nasty merge conflicts with one of my following branches.

Kharonus committed:
[#42158] fixed PR issues

- replaced files color classes with file type classes
- extracted some service code to function higher order
- some css maintenance

HDinger reviewed (changes requested):
> * we can add more mimetype mappings, should be very simple, which is why I extracted that into a factory function. Any suggestions which ones to add? Or should I asked product team?

Yes, I'd ask the product team. I just stumbled over these two examples (.ppt and .xls) as they were part of the figma example.

> * I thought I hit the default button pretty well, at least, it was specified in figma like here: https://www.figma.com/file/gtLQfPe09X7XugAH8L7dTy/Nextcloud-Integration?node-id=936%3A32804. Do you think, we need more styles on it? EDIT: rechecked it, there were some spacings missing, due do the inner spot-icon having no width and height set. fixed.

Currently, the button still does not look as specified, because there is a dark border around it. I think you were having these issues because you are mixing spot styles in with the old button style here. However, as for buttons, we currently stick completely to the old styles. The button designed in figma is our default button so normally you'd need no adjustments at all. Your html should look somehow like this:
```
<button class="button">
  <op-icon icon-classes="button--icon spot-icon spot-icon_external-link" class="op-icon--wrapper">
    <i  class="button--icon spot-icon spot-icon_external-link"></i>
  </op-icon>
  <span class="button--text">Open Nextcloud</span>
</button>
```
All additional styling rules can then be omitted. If the button looks different than specified with that code, it is a specification error imho as it was an explicit decision to go for consistency in the button styles for now. The code above delivers exactly that. You may ask the product team about that as well.

Kharonus committed:
[#42158] replace custom button with standard button

- extended file icon mappings

Kharonus commented:
Switched the button and waiting for an answer from @psatyal for the mimetype mapping

Kharonus committed:
Merge branch 'dev' into implementation/42158-add-direct-action-hover-menu

HDinger reviewed (commented):
Looks good so far. I only have one minor issue left:
The button has too much space to the right:
<img width="524" alt="Bildschirmfoto 2022-06-16 um 07 53 54" src="https://user-images.githubusercontent.com/7457313/174001907-fc07e338-27d0-46ea-8a4d-a9e3912ccdd2.png">

This is due to our badly designed old button styles which add a default margin to every button. You can use the class `button_no-margin` to avoid that.

Kharonus committed:
[#42158] added no margin overwrite

HDinger reviewed (approved):
👍🏻 thanks for the patience during the many iterations! It's good to merge from my side 👍🏻
